helix-core/src/main/java/org/apache/helix/manager/zk/CallbackHandler.java

```diff
@@ -200,6 +208,37 @@ protected void handleEvent(NotificationContext event) {
         logger.warn("Exception in callback processing thread. Skipping callback", e);
       }
     }
+
+    public boolean queueIsEmpty() {
+      return _eventQueue.isEmpty();
+    }
+  }
+
+  class RefreshTask implements Runnable {
+    @Override
+    public void run() {
+      try {
+        while (true) {
+          long currentTime = System.currentTimeMillis();
+          long remainingTime = _lastEventTime + _periodicRefreshTriggerInterval - currentTime;
+          if (remainingTime <= 0) {
+            // If there is no event in the queue, meaning this long idle time could be due to lack of events
+            // Otherwise, if there is event in the queue, this long idle time is due to slow process of events
+            if (_batchCallbackProcessor == null || _batchCallbackProcessor.queueIsEmpty()) {
+              NotificationContext changeContext = new NotificationContext(_manager);
+              changeContext.setType(Type.CALLBACK);
+              changeContext.setChangeType(_changeType);
+              enqueueTask(changeContext);
+            }
+            break;
+          } else {
+            Thread.sleep(remainingTime);
```

Comment:
I know that you are not holding any lock here, but I would prefer using wait(remainingTime)